models/User.js

```diff
@@ -0,0 +1,56 @@
+const mongoose = require('mongoose');
+const { Schema } = mongoose;
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+
+const UserSchema = new Schema({
+  name: {
+    type: String,
+    required: [true, 'Please provide a name'],
+    minLength: 3,
+    maxLength: 50,
+  },
+  email: {
+    type: String,
+    required: [true, 'Please provide an email'],
+    unique: true,
+    match: [
+      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
+      ,
+      'Please provide valid email',
+    ],
+  },
+  password: {
+    type: String,
+    required: [true, 'Please provide a password'],
+    minLength: 6,
+  },
+});
+
+//using mongoose middleware hashing password
+UserSchema.pre('save', async function () {
+  //generate random bytes using method genSalt() with number of rounds
+  const salt = await bcrypt.genSalt(10);
+  // ``this``` in this fn will to user document that's why we use regular function instead of arrow function
+  // because this in arrow function will refer to the window object in the browser or the global object in Node.js
+  // hash password using method hash() with password and salt in order to safely to store in our db
+  this.password = await bcrypt.hash(this.password, salt);
+});
+
+//using mongoose middleware to generate token
+UserSchema.methods.generateJWT = function () {
+  //generate token using method sign() with payload and secret key
+  return jwt.sign(
+    { userId: this._id, name: this.name },
+    process.env.JWT_SECRET,
+    {
+      expiresIn: process.env.JWT_LIFETIME,
+    }
+  );
+};
+
+UserSchema.methods.comparePassword = async function (candidatePassword) {
+  const isMatch = await bcrypt.compare(candidatePassword, this.password);
```

Comment:
For some extra fun facts, lookup what a timing attack is, and why we can't use `===` to compare password hashes securely!

Reply:
A timing attack is a technique used by hackers to try to guess a secret, such as a password or encryption key, by measuring the time it takes a computer or system to respond to different inputs. The hacker can learn more about the secret they are trying to guess by observing how long it takes for the system to respond to each input. For example, if a password is correct, it might take a bit longer for the system to respond than if the password is wrong. A timing attack takes advantage of the fact that some operations take longer to execute than others, allowing the hacker to use the timing information to deduce the secret.

Comparing two password hashes using the ```===``` operator can leak timing information that an attacker could use to try to figure out the correct password. They could do this by measuring the time it takes for the comparison process to complete and making guesses based on that timing data. To prevent this kind of attack, it's better to use a constant-time comparison method that always takes the same amount of time, no matter what the values being compared are. Many libraries, such as the ```bcrypt``` library, provide a ```compare``` method that is always ```constant-time``` and can be used to compare password hashes securely.